### ISSUE TITLE
Softmax function speed, for different batch sizes 

### ISSUE THREAD

**Issue by danyaljj** [pytorch/pytorch#64236]
## A potential  🐛 bug 

I see a significant slowdown in [the softmax function](https://pytorch.org/docs/stable/generated/torch.nn.Softmax.html), as the number of batches. 

## To Reproduce

```python 
import torch
from contextlib import contextmanager
import time


@contextmanager
def _log_time_usage(prefix=""):
    '''log the time usage in a code block
    prefix: the prefix text to show
    '''
    start = time.time()
    try:
        yield
    finally:
        end = time.time()
        elapsed_seconds = float("%.2f" % (end - start))
        print(f'batch={prefix}: elapsed seconds: {elapsed_seconds}')


def compute_and_time_softmax2(tensor, batch_size, repeat):
    with _log_time_usage(batch_size):
        # repeat so that the execution time is more reliable
        for _ in range(repeat):
            torch.nn.Softmax(dim=-1)(tensor)


device = 'cuda'

for batch_size in [1, 4, 16, 64, 256, 1024]:
    compute_and_time_softmax2(torch.randn([batch_size, 1, 50257]).to(device), batch_size, repeat=10000)

# when the input tensors are large, softmax quickly slows down
# batch= 1: elapsed seconds: 0.49
# batch=4: elapsed seconds: 0.36
# batch=16: elapsed seconds: 0.55
# batch=64: elapsed seconds: 2.05
# batch=256: elapsed seconds: 8.25
# batch=1024: elapsed seconds: 31.92
```

## Expected behavior

I was expecting the softmax durations to remain almost flat, as I increase the number of batch sizes.

## Environment

```
Collecting environment information...
PyTorch version: 1.7.1+cu101
Is debug build: False
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.10

Python version: 3.7.10 (default, Feb 26 2021, 18:47:35)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-1049-gcp-x86_64-with-debian-buster-sid
Is CUDA available: True
CUDA runtime version: 9.1.85
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 460.91.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.20.3
[pip3] torch==1.7.1+cu101
[pip3] torchaudio==0.7.2
[pip3] torchvision==0.8.2+cu101
[conda] Could not collect
```

FYI @wellecks

cc @VitalyFedyunin @ngimel @heitorschueroff @albanD @mruberry @jbschlosser

**Comment by ngimel** [comment]
Why would you expect softmax duration to not change as it processes more and more data?

**Comment by danyaljj** [comment]
My expectation was that the **batches are processed in parallel** and hence, more batches should have a significant effect on the run time. 

Note that this does happen when the tensors are small; if I run the same code on tensors of smaller sizes (`[batch x 1 x 128]` instead of `[batch x 1 x 50k]`) we don't see any performance degradation as we increase the number of batches. .

```
for batch_size in [1, 4, 16, 64, 256, 1024]:
    compute_and_time_softmax2(torch.randn([batch_size, 1, 128]).to(device), batch_size, repeat=50000)

# when the tensors are small (dimensions: batch x 1 x 128), the softmax function easily scales to larger batches 
# 1: elapsed seconds: 1.37
# 4: elapsed seconds: 1.4
# 16: elapsed seconds: 1.37
# 64: elapsed seconds: 1.36
# 256: elapsed seconds: 1.37
# 1024: elapsed seconds: 1.36
```

**Comment by ngimel** [comment]
You are not measuring runtime correctly (you should use torch.utils.benchmark.Timer), and also, parallel execution doesn't help when you saturate memory bandwidth. The proper metric for execution efficiency is achieved effective memory bandwidth, so it's expected that larger batches will take longer to process.

**Comment by danyaljj** [comment]
Thanks for the tips! 
I did another timing experiment with PyTorch's benchmarking timer and observed similar results: 

```python 
import torch.utils.benchmark as benchmark

def softmax(x):
    torch.nn.Softmax(dim=-1)(x)

dim = 128
for batch_size in [1, 4, 16, 64, 256, 1024]:
    print(" - - - - - ")
    x = torch.randn([batch_size, 1, dim]).to(device)
    t0 = benchmark.Timer(
        stmt='softmax(x)',
        setup='from __main__ import softmax',
        globals={'x': x})
    print(f'batch-size= {batch_size}, x.size(): {x.size()} -> time for softmax(x):  {t0.timeit(100)}')

# here the run times remain flat: 
# batch-size= 1, x.size(): torch.Size([1, 1, 128]) -> time for softmax(x):  <torch.utils.benchmark.utils.common.Measurement object at 0x7f111bc21610>
# softmax(x)
#   26.51 us
#   1 measurement, 100 runs , 1 thread
#  - - - - - 
# batch-size= 4, x.size(): torch.Size([4, 1, 128]) -> time for softmax(x):  <torch.utils.benchmark.utils.common.Measurement object at 0x7f111bc21510>
# softmax(x)
#   26.74 us
#   1 measurement, 100 runs , 1 thread
#  - - - - - 
# batch-size= 16, x.size(): torch.Size([16, 1, 128]) -> time for softmax(x):  <torch.utils.benchmark.utils.common.Measurement object at 0x7f111bc21610>
# softmax(x)
#   27.07 us
#   1 measurement, 100 runs , 1 thread
#  - - - - - 
# batch-size= 64, x.size(): torch.Size([64, 1, 128]) -> time for softmax(x):  <torch.utils.benchmark.utils.common.Measurement object at 0x7f111bc21510>
# softmax(x)
#   28.33 us
#   1 measurement, 100 runs , 1 thread
#  - - - - - 
# batch-size= 256, x.size(): torch.Size([256, 1, 128]) -> time for softmax(x):  <torch.utils.benchmark.utils.common.Measurement object at 0x7f111bc21610>
# softmax(x)
#   28.91 us
#   1 measurement, 100 runs , 1 thread
#  - - - - - 
# batch-size= 1024, x.size(): torch.Size([1024, 1, 128]) -> time for softmax(x):  <torch.utils.benchmark.utils.common.Measurement object at 0x7f111bc21510>
# softmax(x)
#   28.20 us
#   1 measurement, 100 runs , 1 thread



dim = 50257
for batch_size in [1, 4, 16, 64, 256, 1024]:
    print(" - - - - - ")
    x = torch.randn([batch_size, 1, dim]).to(device)
    t0 = benchmark.Timer(
        stmt='softmax(x)',
        setup='from __main__ import softmax',
        globals={'x': x})
    print(f'batch-size= {batch_size}, x.size(): {x.size()} -> time for softmax(x): {t0.timeit(100)}')

# here run times quickly increase 
# batch-size= 1, x.size(): torch.Size([1, 1, 50257]) -> time for softmax(x): <torch.utils.benchmark.utils.common.Measurement object at 0x7f111bc21610>
# softmax(x)
#   102.26 us
#   1 measurement, 100 runs , 1 thread
#  - - - - - 
# batch-size= 4, x.size(): torch.Size([4, 1, 50257]) -> time for softmax(x): <torch.utils.benchmark.utils.common.Measurement object at 0x7f111bc21510>
# softmax(x)
#   103.18 us
#   1 measurement, 100 runs , 1 thread
#  - - - - - 
# batch-size= 16, x.size(): torch.Size([16, 1, 50257]) -> time for softmax(x): <torch.utils.benchmark.utils.common.Measurement object at 0x7f111bc21610>
# softmax(x)
#   122.07 us
#   1 measurement, 100 runs , 1 thread
#  - - - - - 
# batch-size= 64, x.size(): torch.Size([64, 1, 50257]) -> time for softmax(x): <torch.utils.benchmark.utils.common.Measurement object at 0x7f111bc21510>
# softmax(x)
#   242.17 us
#   1 measurement, 100 runs , 1 thread
#  - - - - - 
# batch-size= 256, x.size(): torch.Size([256, 1, 50257]) -> time for softmax(x): <torch.utils.benchmark.utils.common.Measurement object at 0x7f111bc21610>
# softmax(x)
#   966.73 us
#   1 measurement, 100 runs , 1 thread
#  - - - - - 
# batch-size= 1024, x.size(): torch.Size([1024, 1, 50257]) -> time for softmax(x): <torch.utils.benchmark.utils.common.Measurement object at 0x7f111bc21510>
# softmax(x)
#   3.55 ms
#   1 measurement, 100 runs , 1 thread
```

I suppose this is due to what you hinted at in your easier response that 
>  parallel execution doesn't help when you saturate memory bandwidth 
i.e., when the tensors already large, there is not much room (memory) left to do parallel processing.
